### PR TITLE
Add HEIC encode judgment, and bugfix for a memory leak

### PR DESCRIPTION
**Pull request by LululuSir**
### New Pull Request Checklist

* [ ] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ ] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ ] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [ ] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [ ] I have added the required tests to prove the fix/feature I am adding
* [ ] I have updated the documentation (if necessary)
* [ ] I have run the tests and they pass
* [ ] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / reffers to the following issues: ...

### Pull Request Description

...

### CODE REVIEW

**Comment by LululuSir** [comment]
I found that, HEIC decode will invalid in iOS simulator. We should return NO when the API can't decode a file format, instead of return YES but do nothing ,just return nil when others call "decodedImageWithData:"

**Comment by LululuSir** [comment]
Sorry , I ignore other platforms such as tvos, mac.

File: SDWebImage/SDWebImageImageIOCoder.m [hunk 1]
```diff
@@ -484,9 +497,9 @@ + (BOOL)canEncodeToHEICFormat {
             canEncode = NO;
         } else {
             // Can encode to HEIC
-            CFRelease(imageDestination);
             canEncode = YES;
         }
+        CFRelease(imageDestination);
```
Comment (dreampiggy) [inline SDWebImage/SDWebImageImageIOCoder.m]: ...Wait, why you want to call `CFRelease` on a `NULL` ?

**Comment by dreampiggy** [comment]
@LululuSir HEIF is available on iOS 11(not simluator), macOS 10.13. Why currently we do not add this check is because in most cases, `SDWebImageImageIOCoder` is the last one in coders manager, which is the base coder. So even you return `NO`, this can have no help for image decoding.

But to say, if you move ImageIO coder to have a higher priority, and you have custom HEIC coder(really ? looks great, could you please provide this ?), this may be a problem. This version check seems work but not really good(hard-code). Maybe we can find another more common case like that thing on `canEncodedToHEIC`

**Comment by dreampiggy** [comment]
See #2146

**Comment by dreampiggy** [comment]
Merged by #2146

File: SDWebImage/SDWebImageImageIOCoder.m (hunk 1 above)
Comment (LululuSir) [reply SDWebImage/SDWebImageImageIOCoder.m]: My misstake,ignore it.

**Comment by LululuSir** [comment]
For a user, we should know if we can decode HEIF, before we request data via network. Such as:
`
    NSString *urlPath = @"http://***";
    NSString *targetUrlString;
    if ([[SDWebImageCodersManager sharedInstance] canDecodeFromat:SDImageFormatHEIC]) {
        targetUrlString = [NSString stringWithFormat:@"%@/hello.heic", urlPath];
    } else {
        targetUrlString = [NSString stringWithFormat:@"%@/hello.png", urlPath];
    }
    [[SDWebImageDownloader sharedDownloader] downloadImageWithURL:[NSURL URLWithString:targetUrlString]
                                                          options:SDWebImageDownloaderLowPriority
                                                         progress:NULL
                                                        completed:NULL];
`

**Comment by dreampiggy** [comment]
@LululuSir This check is easy to write and you can check it yourself outside our framework.


And to say, this behavior is based on current system requirement and may chaned in the future. So I think it's not suitable to expose public or add any logic related to this.

**Pull request closed without merging**